5_Subtree.cpp

```diff
@@ -0,0 +1,229 @@
+/*
+ Microsoft Interview Coaching Pilot Program MICP 2018
+Author: Akshama
+Language: C++
+
+/**********WEEK FIVE***********/
+/*
+Given two non-empty binary trees s and t, check whether tree t has exactly the same structure and node values
+with a subtree of s. A subtree of s is a tree consists of a node in s and all of this node's descendants.
+The tree s could also be considered as a subtree of itself.
+
+//TEBOW IT
+
+/*
+1. T - TALK
+Input: Two node pointers, one is the root of tree s and the other is a root of tree t.
+Output: True if tree t has exactly the same structure and node values with a subtree of s, false otherwise.
+
+Clarified:
+Duplicate values in tree: May be there.
+Are the trees BST's (Binary Search Trees): No
+Handling NULL input: An empty tree is a subtree of an tree, return true if t is NULL.
+
+2. E - EXAMPLES
+Example 1:
+Given tree s:
+
+     3
+    / \
+   4   5
+  / \
+ 1   2
+
+Given tree t:
+   4
+  / \
+ 1   2
+Return true, because t has the same structure and node values with a subtree of s.
+
+Example 2:
+Given tree s:
+
+     3
+    / \
+   4   5
+  / \
+ 1   2
+    /
+   0
+Given tree t:
+   4
+  / \
+ 1   2
+Return false.
+
+Example 3:
+Given tree s: NULL
+Given tree t: NULL
+Return true.
+
+Example 4:
+Given tree s: 1
+Given tree t: NULL
+Return true.
+
+Example 3:
+Given tree s: NULL
+Given tree t: 1
+Return false.
+
+3. B - BRUTE-FORCE
+
+Carry out Pre-order traversal of both the trees, t and s. Create a string representation from the traversal.
+(Take X as special character insertion for every NULL node.) If the string representation of t's traversal matches with a
+substring of s's traversal string, then t is a subtree of s.
+Complexity:
+Time:   O(n+m)
+Space:  O(n+m)
+n,m are the number of nodes in s and t respectively.
+
+4. O - OPTIMISE

```

Comment:
Great to see this introduction! What is the time and space complexity in this case?